cpp/src/rolling/lead_lag_nested_detail.cuh

```diff
@@ -0,0 +1,222 @@
+/*
+ * Copyright (c) 2021, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <cudf/column/column.hpp>
+#include <cudf/column/column_view.hpp>
+#include <cudf/copying.hpp>
+#include <cudf/detail/gather.cuh>
+#include <cudf/detail/scatter.cuh>
+#include <cudf/utilities/traits.hpp>
+#include <vector>
+
+namespace cudf::detail {
+namespace {
+/**
+ * @brief Functor to calculate the gather map used for calculating LEAD/LAG.
+ *
+ * @tparam op Aggregation Kind (LEAD vs LAG)
+ * @tparam PrecedingIterator Iterator to retrieve preceding window bounds
+ * @tparam FollowingIterator Iterator to retrieve following window bounds
+ */
+template <aggregation::Kind op, typename PrecedingIterator, typename FollowingIterator>
+struct lead_lag_gather_map_builder {
+  lead_lag_gather_map_builder(size_type input_size,
+                              size_type row_offset,
+                              PrecedingIterator preceding,
+                              FollowingIterator following)
+    : _input_size{input_size},
+      NULL_INDEX{input_size + 1},
+      _row_offset{row_offset},
+      _preceding{preceding},
+      _following{following}
+  {
+  }
+
+  template <aggregation::Kind o = op, CUDF_ENABLE_IF(o == aggregation::LEAD)>
+  size_type __device__ operator()(size_type i)
+  {
+    // Note: grouped_*rolling_window() trims preceding/following to
+    // the beginning/end of the group. `rolling_window()` does not.
+    // Must trim _following[i] so as not to go past the column end.
+    auto following = min(_following[i], _input_size - i - 1);
+    return (_row_offset > following) ? NULL_INDEX : (i + _row_offset);
+  }
+
+  template <aggregation::Kind o = op, CUDF_ENABLE_IF(o == aggregation::LAG)>
+  size_type __device__ operator()(size_type i)
+  {
+    // Note: grouped_*rolling_window() trims preceding/following to
+    // the beginning/end of the group. `rolling_window()` does not.
+    // Must trim _preceding[i] so as not to go past the column start.
+    auto preceding = min(_preceding[i], i + 1);
+    return (_row_offset > (preceding - 1)) ? NULL_INDEX : (i - _row_offset);
+  }
+
+ private:
+  size_type _input_size;
+  size_type NULL_INDEX;
```

Comment:
I don't like using `ALL_CAP` for variable name :(

In addition, for all variables that will not be changed after initialization, then please add `const` qualifier.